test/failure-agent/microshift-failure-agent.sh

```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+
+set -xeuo pipefail
+
+GREENBOOT_CONFIGURATION_FILE=/etc/greenboot/greenboot.conf
+
+PARAM="microshift_failure_scenarios"
+
+CLEANUP_CMDS=()
+
+cleanup() {
+    for cmd in "${CLEANUP_CMDS[@]}"; do
+        ${cmd}
+    done
+}
+trap "cleanup" SIGTERM SIGQUIT INT
+
+get_cmdline_value() {
+    local -r key="${1}"
+
+    if ! grep -q "${PARAM}" /proc/cmdline; then
+        echo ""
+        return
+    fi
+    sed -e "s/^.*${key}=//" -e 's/ .*$//' /proc/cmdline
+}
+
+prevent_backup() {
+    local -r path="/var/lib/microshift-backups"
+    if [[ ! -e /var/lib/microshift-backups ]]; then
+        mkdir -p /var/lib/microshift-backups
+    fi
+    # prevents from creating a new backup directory, but doesn't prevent from updating health.json
+    chattr +i /var/lib/microshift-backups
+    CLEANUP_CMDS+=("chattr -i /var/lib/microshift-backups")
+}
+
+fail_greenboot() {
+    local -r path="/etc/greenboot/check/required.d/99_microshift_test_failure.sh"
+    cat >"${path}" <<EOF
+#!/bin/bash
+false
+EOF
+    CLEANUP_CMDS+=("rm ${path}")
+}
+
+if ! /usr/bin/grub2-editenv list | grep boot_counter; then
+    echo "boot_counter missing - script only for newly staged deployments"
+    exit 0
+fi
+
+boot_counter=$(/usr/bin/grub2-editenv list | grep boot_counter | sed 's,boot_counter=,,')
+
+if test -f "${GREENBOOT_CONFIGURATION_FILE}"; then
+    # shellcheck source=/dev/null
+    source "${GREENBOOT_CONFIGURATION_FILE}"
+fi
+
+if [ -v GREENBOOT_MAX_BOOT_ATTEMPTS ]; then
+    max_boot_attempts="${GREENBOOT_MAX_BOOT_ATTEMPTS}"
+else
+    max_boot_attempts=3
+fi
+CURRENT_BOOT=$((max_boot_attempts - boot_counter + 1))
+
+scenarios_param=$(get_cmdline_value "${PARAM}")
+IFS=',' read -ra scenarios <<<"${scenarios_param}"
+
+for scenario_request in "${scenarios[@]}"; do
+    scenario=${scenario_request%:*}
+    boot=${scenario_request#*:}
+    if [[ "${boot}" == "every" ]] || [[ "${boot}" -eq "${CURRENT_BOOT}" ]]; then
+        ${scenario}
+    fi
+done
+
+sleep infinity
```

Comment:
Is this because it is a service? Is there not a way to specify that something runs once and exits?

Reply:
I don't want it to exit.
For example, I want to simulate that making a backup on first boot of new deployment fails, but I want it to succeed on second boot (greenboot's reboot) - for this reason, I want this service to last up until the shutdown and revert `chattr` - otherwise it'll be present in rolled back deployment as well

Reply:
OK, I see what you're going for. I'm a little worried about relying on a clean exit handler, but let's see how it works in practice. If we find issues we can look at options like persisting the instructions for each boot separately.